test/e2e/network/network_policy.go

```diff
@@ -1874,87 +1874,115 @@ var _ = SIGDescribe("NetworkPolicy [Feature:SCTPConnectivity][LinuxOnly][Disrupt
 			})
 		})
 	})
 })
 
-func testCanConnect(f *framework.Framework, ns *v1.Namespace, podName string, service *v1.Service, targetPort int) {
-	testCanConnectProtocol(f, ns, podName, service, targetPort, v1.ProtocolTCP)
+func testCanConnect(f *framework.Framework, ns *v1.Namespace, podName string, service *v1.Service, targetPort int, tryCount int) {
+	testCanConnectProtocol(f, ns, podName, service, targetPort, v1.ProtocolTCP, tryCount)
 }
 
-func testCannotConnect(f *framework.Framework, ns *v1.Namespace, podName string, service *v1.Service, targetPort int) {
-	testCannotConnectProtocol(f, ns, podName, service, targetPort, v1.ProtocolTCP)
+func testCannotConnect(f *framework.Framework, ns *v1.Namespace, podName string, service *v1.Service, targetPort int, tryCount int) {
+	testCannotConnectProtocol(f, ns, podName, service, targetPort, v1.ProtocolTCP, tryCount)
 }
 
-func testCanConnectProtocol(f *framework.Framework, ns *v1.Namespace, podName string, service *v1.Service, targetPort int, protocol v1.Protocol) {
-	ginkgo.By(fmt.Sprintf("Creating client pod %s that should successfully connect to %s.", podName, service.Name))
-	podClient := createNetworkClientPod(f, ns, podName, service, targetPort, protocol)
-	defer func() {
-		ginkgo.By(fmt.Sprintf("Cleaning up the pod %s", podClient.Name))
-		if err := f.ClientSet.CoreV1().Pods(ns.Name).Delete(context.TODO(), podClient.Name, metav1.DeleteOptions{}); err != nil {
-			framework.Failf("unable to cleanup pod %v: %v", podClient.Name, err)
+func testCanConnectProtocol(f *framework.Framework, ns *v1.Namespace, podName string, service *v1.Service, targetPort int, protocol v1.Protocol, tryCount int) {
+	i := 0
+	keepTrying := true
+	var podClient *v1.Pod
+	var conErr error
+
+	// Network policy can take some time to program, especially in scaled/busy environments. Let's allow some retries to
+	// account for that, until we have status field available on network policy object
+	for i < tryCount && keepTrying {
+		ginkgo.By(fmt.Sprintf("Creating client pod %s that should successfully connect to %s.", podName, service.Name))
+		framework.Logf(fmt.Sprintf("Count: %d of tryCount: %d - Creating client pod %s that should successfully connect to %s.", i+1, tryCount, podName, service.Name))
+		podClient = createNetworkClientPod(f, ns, podName, service, targetPort, protocol)
+		conErr = checkConnectivity(f, ns, podClient, service)
+		if conErr == nil {
+			// got the expected result, stop trying
+			keepTrying = false
 		}
-	}()
-	checkConnectivity(f, ns, podClient, service)
-}
 
-func testCannotConnectProtocol(f *framework.Framework, ns *v1.Namespace, podName string, service *v1.Service, targetPort int, protocol v1.Protocol) {
-	ginkgo.By(fmt.Sprintf("Creating client pod %s that should not be able to connect to %s.", podName, service.Name))
-	podClient := createNetworkClientPod(f, ns, podName, service, targetPort, protocol)
-	defer func() {
 		ginkgo.By(fmt.Sprintf("Cleaning up the pod %s", podClient.Name))
```

Comment:
This won't work; in the original function, the cleanup is `defer`ed, so it happens after the debug `collectPodsAndNetworkPolicies` step, if that happens. With the code here, the pod will be deleted before you collect its logs. (But you have to be careful with variable scope if you `defer` _all_ of the cleanup...)

Reply:
Ah, nice catch. I will fix this and push an update.